sklearn/tree/tests/test_monotonic_tree.py

```diff
@@ -0,0 +1,348 @@
+        )
+        with pytest.raises(ValueError, match=msg):
+            est.fit(X, y)
+
+
+def assert_1d_reg_tree_children_monotonic_bounded(tree_, monotonic_sign):
+    # Flip values to always check for increasing constraint
+    values = monotonic_sign * tree_.value
+    for i in range(tree_.node_count):
+        if tree_.children_left[i] > i and tree_.children_right[i] > i:
+            # Check monotonicity on children
+            i_left = tree_.children_left[i]
+            i_right = tree_.children_right[i]
+            assert values[i_left] <= values[i_right]
+            val_middle = values[i]
+            # Check bounds on grand-children, filtering out leaf nodes
+            if tree_.feature[i_left] >= 0:
+                i_left_right = tree_.children_right[i_left]
+                assert values[i_left_right] <= val_middle
+            if tree_.feature[i_right] >= 0:
+                i_right_left = tree_.children_left[i_right]
+                assert val_middle <= values[i_right_left]
+
+
+def assert_1d_reg_monotonic(clf, monotonic_sign, min_x, max_x, n_steps):
+    X_grid = np.linspace(min_x, max_x, n_steps).reshape(-1, 1)
+    y_pred_grid = clf.predict(X_grid)
+    assert (monotonic_sign * np.diff(y_pred_grid) >= 0.0).all()
+
+
+@pytest.mark.parametrize("monotonic_sign", (-1, 1))
+@pytest.mark.parametrize("splitter", ("best", "random"))
+@pytest.mark.parametrize("depth_first", (True, False))
+def test_1d_tree_nodes_values(
+    monotonic_sign, splitter, depth_first, global_random_seed
+):
+    # Adaptation from test_nodes_values in test_montonic_constraints.py
+    # Build a single tree with only one feature, and make sure the nodes
+    # values respect the monotonic constraints.
+
+    # Considering the following tree with a monotonic POS constraint, we
+    # should have:
+    #
+    #       root
+    #      /    \
+    #     a      b
+    #    / \    / \
+    #   c   d  e   f
+    #
+    # c <= d <= root <= e <= f
+
+    rng = np.random.RandomState(global_random_seed)
+    n_samples = 1000
+    n_features = 1
+    X = rng.rand(n_samples, n_features)
+    y = rng.rand(n_samples)
+
+    if depth_first:
+        # No max_leaf_nodes, default depth first tree builder
+        clf = DecisionTreeRegressor(
+            splitter=splitter,
+            monotonic_cst=[monotonic_sign],
+            random_state=global_random_seed,
+        )
+    else:
+        # max_leaf_nodes triggers best first tree builder
+        clf = DecisionTreeRegressor(
+            splitter=splitter,
+            monotonic_cst=[monotonic_sign],
+            max_leaf_nodes=n_samples,
+            random_state=global_random_seed,
+        )
+    clf.fit(X, y)
+
+    assert_1d_reg_tree_children_monotonic_bounded(clf.tree_, monotonic_sign)
+    assert_1d_reg_monotonic(clf, monotonic_sign, np.min(X), np.max(X), 100)
+
+
+def assert_nd_reg_tree_children_monotonic_bounded(tree_, monotonic_cst):
```

Comment:
Again, and ideally, this function must be tested to raise.

Reply:
I am working on this.

Reply:
This was addressed in `test_assert_nd_reg_tree_children_monotonic_bounded`.